charts/nginx-ingress-services/templates/certificate.yaml

```diff
@@ -0,0 +1,40 @@
+{{- if and .Values.tls.enabled .Values.tls.useCertificateManager -}}
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: "{{ include "nginx-ingress-services.zone" . | replace "." "-" }}-csr"
+  namespace: {{ .Release.Namespace }}
+  docs: "https://cert-manager.io/docs/usage/certificate"
+  labels:
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  issuerRef:
+    name: letsencrypt-http01
+    kind: ClusterIssuer
+  usages:
+    - server auth
+  duration: 2160h     # 90d; NOTE: ignored by Letsencrypt
+  renewBefore: 360h   # 15d
+  isCA: false
+  keyAlgorithm: ecdsa
+  keySize: 384        # 521 is not supported by Letsencrypt
+  keyEncoding: pkcs1
+  secretName: {{ include "nginx-ingress-services.getCertificateSecretName" . | quote }}
+  # NOTE: disabled due to https://github.com/jetstack/cert-manager/issues/2978
+  # TODO: enable when fixed (probably when cert-manager:v0.16 released)
+  #privateKey:
+  #  rotationPolicy: Always
+  dnsNames:
+    - {{ .Values.config.dns.https }}
+    - {{ .Values.config.dns.ssl }}
+    - {{ .Values.config.dns.webapp }}
+    - {{ .Values.config.dns.fakeS3 }}
+    {{- if .Values.teamSettings.enabled }}
+    - {{ .Values.config.dns.teamSettings }}
+    {{- end }}
+    {{- if .Values.accountPages.enabled }}
+    - {{ .Values.config.dns.accountPages }}
+    {{- end }}
```

Comment:
Isn't it better to use annotation on every service than having one certificate for all?
That way we don't have to maintain this certificate file and we wouldn't have to regenerate certificates when we change this file.

Reply:
I would agree to the part where we should question whether we want to issue dedicated certificates for each FQDN. I'm not sure. What's the thread model? What could go wrong?

I disagree with the annotation approach. Personally, I find this way of separating resources somewhat cleaner. And, if I'm not mistaken, a dedicated certificate file gives us more control over certificate parameters. But maybe I'm wrong. Can you elaborate a bit? Maybe describe the different approaches in a use case to show the advantage of annotating a ingress resource?

Reply:
I wasn't talking about any threats here. I just meant to say that we don't need to manage another resource here. We can just add annotations to the ingresses and get the certificates to work. As an additional benefit, we get separate certificates for separate services.

About cleanliness of separation, I suppose we have a choice between separating certificates from ingresses and separating certificates for different services. I like to think certificates as part of the ingress rather than separate from them and think of services as separate from each other. When we make one certificate for all services, we are tying them together. And so when we create a new service, it is not be the most obvious thing to find the certificate for all services and another domain there.

> a dedicated certificate file gives us more control over certificate parameters

This is true, but taking control over things also increases our burden in maintaining them. Maybe we should look at the defaults which cert-manager provides and see if it works for us, I am hopeful that they are secure enough by default. If we don't like it, then of course we should take control over these certificates. But even then, we should see if we can get cert-manager to be better at the defaults and eventually let go of this control so we have less things to do on our plates.

Reply:
A dedicated certificate for each FQDN - I can cope with that.

You are making a valid point about annotations. So, it basically boils down to explicity vs. responsibility. I have to admit, I'm more on the explicit side of things. Furthermore, I dont think, the argument about defaults is not really applicable. With this approach we lose the ability to control the `cert-manager` version. So defaults might change w/o us being aware of that.

I guess its `Certificate`s then? ;) Unless you want to fight me on this?

Reply:
At the end, @akshaymankar and I, we agreed to disagree on annotation vs explicitly defining a `Certificate` resource. I decided to go with the latter.